File: 01-presentation-blog-examples/src/main/java/com/nalepa/demo/example10/Example10_Without_Bulkhead_Pattern.java
package com.nalepa.demo.example10;

import java.time.Duration;
import java.time.LocalTime;
import java.util.LinkedList;

public class Example10_Without_Bulkhead_Pattern {
    public static void main(String[] args) throws InterruptedException {
        LinkedList<Runnable> queue = new LinkedList<>();
        long exampleStart = System.nanoTime();

        queue.add(new MonitoredRunnable(() -> simulateSlowEndpoint(1)));
        queue.add(new MonitoredRunnable(() -> simulateSlowEndpoint(2)));
        queue.add(new MonitoredRunnable(() -> simulateSlowEndpoint(3)));

        queue.add(new MonitoredRunnable(() -> simulateFastEndpoint(1)));
        queue.add(new MonitoredRunnable(() -> simulateFastEndpoint(2)));
        queue.add(new MonitoredRunnable(() -> simulateFastEndpoint(3)));

        Thread workerThread = new Thread(
                () -> {
                    while (!queue.isEmpty()) {
                        Runnable runnable = queue.poll();
                        runnable.run();
                    }
                },
                "worker-thread"
        );
        workerThread.start();
        workerThread.join();
        log("Example ended after: " + Duration.ofNanos(System.nanoTime() - exampleStart).toSeconds() + " s");
    }

    static void simulateSlowEndpoint(int index) {
        log("Start slow for: " + index);
        try {
            Thread.sleep(Duration.ofSeconds(5).toMillis());
        } catch (InterruptedException e) {
            throw new RuntimeException(e);
        }
        log("End slow for: " + index);
    }

    static void simulateFastEndpoint(int index) {
        log("Start fast for: " + index);
        log("End fast for: " + index);
    }

    static void log(String message) {
        System.out.println(LocalTime.now() + " : " + Thread.currentThread().getName() + " : " + message);
    }

    static class MonitoredRunnable implements Runnable {
        private final Runnable delegate;
        private final long runnableInstanceCreatedAt;

        public MonitoredRunnable(Runnable delegate) {
            this.delegate = delegate;
            this.runnableInstanceCreatedAt = System.nanoTime();
        }

        @Override
        public void run() {
            System.out.println();
            log("Queue wait time took: " + Duration.ofNanos(System.nanoTime() - runnableInstanceCreatedAt).toSeconds() + " s");
            long startExecution = System.nanoTime();
            delegate.run();
            log("Task execution took: " + Duration.ofNanos(System.nanoTime() - startExecution).toSeconds() + " s");
        }
    }
}

File: presentation-blog-examples/src/main/kotlin/com/nalepa/demo/example10/Example10_Without_Bulkhead_Pattern.kt
package com.nalepa.demo.example10

import com.nalepa.demo.example12.log
import java.time.Duration
import java.time.LocalTime
import java.util.*

// fast endpoint is waiting in queue :((((((
fun main() {
    val queue = LinkedList<Runnable>()
    val exampleStart = System.nanoTime()

    repeat(4) { index ->
        queue.add(MonitoredRunnable { simulateSlowEndpoint(index) })
    }

    repeat(4) { index ->
        queue.add(MonitoredRunnable { simulateFastEndpoint(index) })
    }

    val workerThread =
        Thread.ofPlatform()
            .name("worker-thread")
            .start {
                while (!queue.isEmpty()) {
                    val runnable = queue.poll()
                    runnable.run()
                }
            }

    workerThread.join()
    log("Example ended after: ${Duration.ofNanos(System.nanoTime() - exampleStart).toSeconds()} s")
}

fun simulateSlowEndpoint(index: Int) {
    log("Start slow for: $index")
    Thread.sleep(Duration.ofSeconds(5))
    log("End slow for: $index")
}

fun simulateFastEndpoint(index: Int) {
    log("Start fast for: $index")
    log("End fast for: $index")
}

fun log(message: String) {
    println("${LocalTime.now()} : ${Thread.currentThread().name} : $message")
}

class MonitoredRunnable(
    private val delegate: Runnable,
) : Runnable {

    val runnableInstanceCreatedAt = System.nanoTime()

    override fun run() {
        log("Queue wait time took: ${Duration.ofNanos(System.nanoTime() - runnableInstanceCreatedAt).toSeconds()} s")

        val startExecution = System.nanoTime()
        delegate.run()
        log("Task execution took: ${Duration.ofNanos(System.nanoTime() - startExecution).toSeconds()} s")
    }
}




